proposals/0000-Move-Platforms-to-Own-Packages.md

```diff
@@ -0,0 +1,277 @@
+---
+title: Move iOS and Android Specific Code to Their Own Packages
+author:
+- Ian Meikle
+date: today
+---
+
+# RFC0000: Move iOS and Android Specific Code to Their Own Packages
+
+## Summary
+
+Re-organize the React Native repo as a Yarn workspace monorepo, and move platform specific JS and native code for iOS and Android into their own packages.
+
+## Basic example
+
+Common APIs and components would remain in `react-native`:
+
+```js
+import {AppRegistry, Alert, View} from 'react-native';
+```
+
+Platform-specific components and APIs would be imported from the platform's module:
+
+```js
+// iOS
+import {Alert} from 'react-native-ios';
+
+// Android
+import {Permissions} from 'react-native-android';
+```
+
+## Motivation
+
+React Native is for more than just iOS and Android - there are renderers for Windows, macOS, and even the DOM and even more on their way.
+
+This is analogous to when React [split into React and ReactDOM](https://reactjs.org/blog/2015/09/10/react-v0.14-rc1.html#two-packages-react-and-react-dom), and (in my opinion at least!) the separation of these modules was a benefit to the renderers of other platforms.
+
+Splitting iOS and Android out into their own packages would make them functionally like out-of-tree platforms. The process of doing this split should hopefully strengthen support for out-of-tree platforms as a result.
+
+The process of re-organizing these modules will, I hope, also result in more components being found that can be split out into their own repo and package as part of the slimmening.
+
+Platform-specific APIs could drop the platform suffix from their name. `PermissionsAndroid` could just be `Permissions` for example.
+
+React Native being a Yarn workspace might make it easier to make other modules that should be developed and released alongside the main project. It would be easier to make a more advanced RNTester app in-tree with its own set of dependencies, for example.
+
+(**UPDATE**: I am working on a pull request to set up a monorepo, so other proposals could benefit from it regardless of this proposal's progress)
```

Comment:
Just fyi: This can probably not be done on GitHub and will have to be done at FB in one go.